sklearn/metrics/_pairwise_distances_reduction.pyx

```diff
@@ -0,0 +1,1512 @@
+        # threads used for the reduction but there won't be allocated but unused
+        # datastructures.
+        self.heaps_proxy_distances_chunks = <DTYPE_t **> malloc(
+            sizeof(DTYPE_t *) * self.effective_omp_n_thread)
+        self.heaps_indices_chunks = <ITYPE_t **> malloc(
+            sizeof(ITYPE_t *) * self.effective_omp_n_thread)
+
+    def __dealloc__(self):
+        if self.heaps_indices_chunks is not NULL:
+            free(self.heaps_indices_chunks)
+
+        if self.heaps_proxy_distances_chunks is not NULL:
+            free(self.heaps_proxy_distances_chunks)
+
+    cdef void _reduce_on_chunks(self,
+        ITYPE_t X_start,
+        ITYPE_t X_end,
+        ITYPE_t Y_start,
+        ITYPE_t Y_end,
+        ITYPE_t thread_num,
+    ) nogil:
+        cdef:
+            ITYPE_t i, j
+            ITYPE_t n_X = X_end - X_start
+            ITYPE_t n_Y = Y_end - Y_start
+            ITYPE_t k = self.k
+            DTYPE_t *heaps_proxy_distances = self.heaps_proxy_distances_chunks[thread_num]
+            ITYPE_t *heaps_indices = self.heaps_indices_chunks[thread_num]
+
+        # Pushing the distance and their associated indices on heaps
+        # which keep tracks of the argkmin.
+        for i in range(n_X):
+            for j in range(n_Y):
+                heap_push(
+                    heaps_proxy_distances + i * self.k,
+                    heaps_indices + i * self.k,
+                    k,
+                    self._datasets_pair.proxy_dist(X_start + i, Y_start + j),
+                    Y_start + j,
+                )
+
+    @final
+    cdef void _on_X_prange_iter_init(self,
+        ITYPE_t thread_num,
+        ITYPE_t X_start,
+        ITYPE_t X_end,
+    ) nogil:
+        # As this strategy is embarrassingly parallel, we can set the
+        # thread heaps pointers to the proper position on the main heaps
+        self.heaps_proxy_distances_chunks[thread_num] = &self.argkmin_distances[X_start, 0]
+        self.heaps_indices_chunks[thread_num] = &self.argkmin_indices[X_start, 0]
+
+    @final
+    cdef void _on_X_prange_iter_finalize(self,
+        ITYPE_t thread_num,
+        ITYPE_t X_start,
+        ITYPE_t X_end,
+    ) nogil:
+        cdef:
+            ITYPE_t idx, jdx
+
+        # Sorting indices of the argkmin for each query vector of X
+        for idx in range(X_end - X_start):
+            simultaneous_sort(
+                self.heaps_proxy_distances_chunks[thread_num] + idx * self.k,
+                self.heaps_indices_chunks[thread_num] + idx * self.k,
+                self.k
+            )
+
+    cdef void _on_Y_init(self,
+        ITYPE_t num_threads,
+    ) nogil:
+        cdef:
+            # Maximum number of scalar elements (the last chunks can be smaller)
+            ITYPE_t heaps_size = self.X_n_samples_chunk * self.k
+            ITYPE_t thread_num
+
+        for thread_num in prange(num_threads, schedule='static', nogil=True,
+                                 num_threads=num_threads):
```

Comment:
Question: is it really useful to pass an explicit prange schedule?

Since the workload is not always homegenous, for instance radius neighbors and argkmin with large k might be quite data-dependent, maybe the `"guided"` schedule might be safer? Also, some modern CPUs such as Apple Silicon M1 have asymmetric cores ([AMP](https://en.wikipedia.org/wiki/Asymmetric_multiprocessing)) so the static schedule might not be optimal for them.

Reply:
The current choice for static scheduling was made on the _a priori_ that the insertion of elements in data-structure is rather cheap compared to distances computations. Under the vectors distribution uncertainty, I preferred favoring `static` which seems to be a robust scheduling initially when implementing `ArgKmin`.

I think this make sense for `ArgKmin` as its the dynamic is rather uniform on threads: `heap_push` is hit every time and not conditionally and because its worst case complexity (O(log(k))) is negligible).

Though, this might not be the case for `RadiusNeighborhood` as some vectors might have a lot of neighbors, but other none. Furthermore element insertion comes with sporadic reallocation via `std::vector::push_back` which can be costly (and potentially blocking depending on the underlying allocator?).

Also, the scheduling algorithms for each strategy (`static`, `guided`, `dynamic`, etc.) depends on each OpenMP runtime implementation. Might some of the, make use of some information regarding the machine architecture to perform the good choice?

A lot of further comparisons can be drawn, [see this comment here](https://github.com/jjerphan/scikit-learn/blob/7e5775c8575a3f2543729fad279ea31c9a452bda/sklearn/metrics/_pairwise_distances_reduction.pyx#L1219-L1223).

In every case, making `schedule` reduction-dependent is easy.

Reply:
iI think dynamic can be problematic by breaking cache alignment. If guided has little overhead in cases where both CPUs are homogeneous and tasks are homogeneous (e.g. argkmin with k==1) then I would be in favor of using it by default default in this PR because it is likely to be more robust when CPUs are randomly loaded with noisy neighbors or not homogeneous (and might work a bit better when tasks are not homogeneous such as radius neighbors).

Reply:
I would suggest studying the impact of the scheduling policy properly in a dedicated PR. WDYT?